File: apps/web/src/lib/chat/contacts.ts
/**
 * Contact detection (phones, @handles, messenger links, emails) — client-side mirror of
 * apps/api/apps/chat/contacts.py for instant feedback. The server is the source of truth;
 * test vectors: apps/api/apps/tests/data/contact_cases.json.
 */

export type ContactKind = "phone" | "handle" | "link" | "email";
export interface ContactHit {
  kind: ContactKind;
  start: number;
  end: number;
}

export const CONTACT_LABELS: Record<ContactKind, string> = {
  phone: "номер телефона",
  handle: "ник",
  link: "ссылку на\u00a0мессенджер",
  email: "почту",
};

const ALLOWED_NUMBERS = new Set(["88002000122", "88003334434", "84950510000"]);

const L = "a-zа-яё";
const DIGIT_WORDS: Record<string, string> = {
  ноль: "0", нуль: "0", один: "1", одна: "1", два: "2", две: "2", три: "3",
  четыре: "4", пять: "5", шесть: "6", семь: "7", восемь: "8", девять: "9",
};
const WORD = Object.keys(DIGIT_WORDS).sort((a, b) => b.length - a.length).join("|");
const UNIT = `(?:\\d|(?<![${L}])(?:${WORD})(?![${L}])|(?<=\\d)[oо]+(?![${L}]))`;
const SEP = "[\\s\\-‐‑–—.()\\[\\]/*_,]{0,3}";
const RUN = new RegExp(`(?:(?:\\+|(?<![${L}])плюс(?![${L}]))\\s*)?${UNIT}(?:${SEP}${UNIT})*`, "giu");
const UNIT_RE = new RegExp(UNIT, "giuy");

const EMAIL = new RegExp(
  `[\\p{L}\\p{N}_.+-]+@[\\p{L}\\p{N}_-]+(?:\\.[\\p{L}\\p{N}_-]+)*\\.[a-zа-я]{2,}` +
    `|[\\p{L}\\p{N}_.+-]{2,}\\s*(?:\\(at\\)|\\[at\\]|\\{at\\}|\\s(?:at|собака)\\s)\\s*[\\p{L}\\p{N}_-]+\\s*` +
    `(?:\\.|\\(dot\\)|\\[dot\\]|\\s(?:dot|точка)\\s)\\s*(?:ru|com|net|org|рф|me|io|su|by|kz|ua|info)(?![a-zа-я])`,
  "giu",
);
const HANDLE = new RegExp(`(?<![\\p{L}\\p{N}_@.])@[a-z][a-z0-9_.]{2,31}(?<!\\.)`, "giu");
const DOMAINS =
  "t\\s*\\.\\s*me|telegram\\.(?:me|org|dog)|wa\\.me|(?:api|chat|web)\\.whatsapp\\.com|whatsapp\\.com" +
  "|(?:m\\.)?vk\\.(?:com|me|cc)|vkontakte\\.ru|ok\\.ru|instagram\\.com|instagr\\.am|facebook\\.com|fb\\.(?:me|com)" +
  "|m\\.me|discord\\.(?:gg|com)|discordapp\\.com|(?:invite\\.)?viber\\.com|signal\\.(?:me|group)|snapchat\\.com" +
  "|tiktok\\.com|twitter\\.com|x\\.com|linkedin\\.com|skype\\.com|join\\.skype\\.com|icq\\.im|max\\.ru|threads\\.net";
const LINK = new RegExp(
  `(?<![\\p{L}\\p{N}_.])(?:https?://)?(?:www\\.)?(?:${DOMAINS})(?:\\s*/\\S*)?|(?:tg|viber|whatsapp|skype|discord)://\\S+`,
  "giu",
);
const MESSENGER =
  "telegram|телеграм\\p{L}*|телег[аеиу]|тг|tg|whats\\s?app|ват?с\\s?ап\\p{L}*|вотс?ап\\p{L}*|вацап\\p{L}*|viber|вайбер\\p{L}*" +
  "|discord|дискорд\\p{L}*|instagram|инстаграм\\p{L}*|инст[аеуы]|insta|vk|вк|вконтакте|signal|skype|скайп\\p{L}*" +
  "|facebook|фейсбук\\p{L}*|snapchat|max";
const INTENT =
  "пиши|напиши|напишите|пишите|написать|черкни|добавь|добавьте|добавляйся|добавляйтесь|стукни|стукните" +
  "|стучи|звони|позвони|позвоните|свяжемся|свяжитесь|связаться|связь|найди|найдите|ищи|ищите|мой|мою|моя" +
  "|мои|контакт\\p{L}*|ник|логин|аккаунт|переписываться|переписка|перейд\\p{L}*|перейти|давай|давайте|есть";
const B = `(?<![${L}\\d_])`;
const E = `(?![${L}\\d_])`;
const NW = "[^\\p{L}\\p{N}_\\n]";
const MENTION = new RegExp(
  `${B}(?:${INTENT})${E}(?:${NW}+[\\p{L}\\p{N}_-]+){0,3}?${NW}+(?:${MESSENGER})${E}` +
    `|${B}(?:${MESSENGER})\\s*[:=]\\s*\\S+` +
    `|${B}(?:${MESSENGER})\\s+[a-z][a-z0-9_.]{3,}`,
  "giu",
);

function toDigits(token: string): string {
  const t = token.toLowerCase();
  return DIGIT_WORDS[t] ?? t.replace(/[oо]/g, "0");
}

type Group = [string, number, number];

function groupsOf(text: string, start: number, end: number): Group[] {
  const groups: Group[] = [];
  let prevEnd = -1;
  let prevWord = true;
  let pos = start;
  const slice = text.slice(0, end);
  while (pos < end) {
    UNIT_RE.lastIndex = pos;
    const m = UNIT_RE.exec(slice);
    if (!m) {
      pos += 1;
      continue;
    }
    const tok = m[0];
    const word = tok.toLowerCase() in DIGIT_WORDS;
    if (groups.length && !word && !prevWord && prevEnd === m.index) {
      const g = groups[groups.length - 1];
      groups[groups.length - 1] = [g[0] + toDigits(tok), g[1], m.index + tok.length];
    } else {
      groups.push([toDigits(tok), m.index, m.index + tok.length]);
    }
    prevEnd = m.index + tok.length;
    prevWord = word;
    pos = prevEnd;
  }
  return groups;
}

function isPhone(digits: string, plusStart: boolean, seps: string, single: boolean): boolean {
  const n = digits.length;
  if (plusStart) return n >= 10 && n <= 15;
  if (n === 11) return "78".includes(digits[0]) && "3489".includes(digits[1]);
  if (n === 10) {
    if (digits[0] === "9") return true;
    return "348".includes(digits[0]) && (single || /[-()]/.test(seps));
  }
  return false;
}

function allowedEnd(groups: Group[], i: number): number | null {
  let digits = "";
  for (let j = i; j < groups.length; j++) {
    digits += groups[j][0];
    if (digits.length > 11) return null;
    if (ALLOWED_NUMBERS.has(digits) || (digits[0] === "7" && ALLOWED_NUMBERS.has("8" + digits.slice(1)))) return j;
  }
  return null;
}

function phoneHits(text: string): ContactHit[] {
  const hits: ContactHit[] = [];
  for (const run of text.matchAll(RUN)) {
    const rs = run.index ?? 0;
    const raw = run[0].toLowerCase();
    const plus = raw.startsWith("+") || raw.startsWith("плюс");
    const groups = groupsOf(text, rs, rs + run[0].length);
    let i = 0;
    while (i < groups.length) {
      const allowed = allowedEnd(groups, i);
      if (allowed !== null) {
        i = allowed + 1;
        continue;
      }
      let digits = "";
      let found: number | null = null;
      for (let j = i; j < groups.length; j++) {
        digits += groups[j][0];
        if (digits.length > 15) break;
        const seps = text.slice(groups[i][2], groups[j][1]);
        if (isPhone(digits, i === 0 && plus, seps, i === j)) {
          found = j;
          break;
        }
      }
      if (found !== null) {
        hits.push({ kind: "phone", start: i === 0 && plus ? rs : groups[i][1], end: groups[found][2] });
        i = found + 1;
      } else i += 1;
    }
  }
  return hits;
}

export function findContacts(text: string): ContactHit[] {
  if (!text) return [];
  const found: ContactHit[] = [];
  const add = (kind: ContactKind, rx: RegExp) => {
    for (const m of text.matchAll(rx)) found.push({ kind, start: m.index ?? 0, end: (m.index ?? 0) + m[0].length });
  };
  add("email", EMAIL);
  add("link", LINK);
  add("handle", HANDLE);
  add("link", MENTION);
  found.push(...phoneHits(text));
  found.sort((a, b) => a.start - b.start || b.end - b.start - (a.end - a.start));
  const out: ContactHit[] = [];
  for (const h of found) {
    if (out.length && h.start < out[out.length - 1].end) continue;
    out.push(h);
  }
  return out;
}

export function describeContacts(hits: ContactHit[]): string {
  return [...new Set(hits.map((h) => CONTACT_LABELS[h.kind]))].join(", ");
}

File: apps/api/apps/chat/contacts.py
"""
Поиск контактов в тексте: телефоны, @ники, ссылки на мессенджеры и соцсети, email.

До первого завершённого созвона клиент и специалист не могут обмениваться контактами
в чате (правило сервиса). Используется для сообщений, названий файлов и текстов
профиля специалиста. Та же логика продублирована на фронтенде
(apps/web/src/lib/chat/contacts.ts) для мгновенной подсказки — источник правды здесь.

Сам текст никогда не логируется — только количество найденных фрагментов.
"""
import re
from dataclasses import dataclass

KIND_LABELS = {
    "phone": "номер телефона",
    "handle": "ник",
    "link": "ссылку на мессенджер",
    "email": "почту",
}

# Телефоны кризисных линий из подсказок сервиса — не личные контакты
ALLOWED_NUMBERS = {"88002000122", "88003334434", "84950510000"}

_L = r"a-zа-яё"  # буквы для «границ слова» (работает одинаково с JS-версией)

_DIGIT_WORDS = {
    "ноль": "0", "нуль": "0", "один": "1", "одна": "1", "два": "2", "две": "2", "три": "3",
    "четыре": "4", "пять": "5", "шесть": "6", "семь": "7", "восемь": "8", "девять": "9",
}
_WORD = "|".join(sorted(_DIGIT_WORDS, key=len, reverse=True))
# Одна «цифра»: 0-9, цифра словом или буква О вместо нуля между цифрами
_UNIT = rf"(?:\d|(?<![{_L}])(?:{_WORD})(?![{_L}])|(?<=\d)[oо]+(?![{_L}]))"
_SEP = r"[\s\-‐‑–—.()\[\]/*_,]{0,3}"
_RUN = re.compile(rf"(?:(?:\+|(?<![{_L}])плюс(?![{_L}]))\s*)?{_UNIT}(?:{_SEP}{_UNIT})*", re.I)
_UNIT_RE = re.compile(_UNIT, re.I)

_EMAIL = re.compile(
    r"[\w.+-]+@[\w-]+(?:\.[\w-]+)*\.[a-zа-я]{2,}"
    r"|[\w.+-]{2,}\s*(?:\(at\)|\[at\]|\{at\}|\s(?:at|собака)\s)\s*[\w-]+\s*"
    r"(?:\.|\(dot\)|\[dot\]|\s(?:dot|точка)\s)\s*(?:ru|com|net|org|рф|me|io|su|by|kz|ua|info)(?![a-zа-я])",
    re.I,
)
_HANDLE = re.compile(r"(?<![\w@.])@[a-z][a-z0-9_.]{2,31}(?<!\.)", re.I)

_DOMAINS = (
    r"t\s*\.\s*me|telegram\.(?:me|org|dog)|wa\.me|(?:api|chat|web)\.whatsapp\.com|whatsapp\.com"
    r"|(?:m\.)?vk\.(?:com|me|cc)|vkontakte\.ru|ok\.ru|instagram\.com|instagr\.am|facebook\.com|fb\.(?:me|com)"
    r"|m\.me|discord\.(?:gg|com)|discordapp\.com|(?:invite\.)?viber\.com|signal\.(?:me|group)|snapchat\.com"
    r"|tiktok\.com|twitter\.com|x\.com|linkedin\.com|skype\.com|join\.skype\.com|icq\.im|max\.ru|threads\.net"
)
_LINK = re.compile(
    rf"(?<![\w.])(?:https?://)?(?:www\.)?(?:{_DOMAINS})(?:\s*/[^\s]*)?"
    r"|(?:tg|viber|whatsapp|skype|discord)://[^\s]+",
    re.I,
)

_MESSENGER = (
    r"telegram|телеграм\w*|телег[аеиу]|тг|tg|whats\s?app|ват?с\s?ап\w*|вотс?ап\w*|вацап\w*|viber|вайбер\w*"
    r"|discord|дискорд\w*|instagram|инстаграм\w*|инст[аеуы]|insta|vk|вк|вконтакте|signal|skype|скайп\w*"
    r"|facebook|фейсбук\w*|snapchat|max"
)
_INTENT = (
    r"пиши|напиши|напишите|пишите|написать|черкни|добавь|добавьте|добавляйся|добавляйтесь|стукни|стукните"
    r"|стучи|звони|позвони|позвоните|свяжемся|свяжитесь|связаться|связь|найди|найдите|ищи|ищите|мой|мою|моя"
    r"|мои|контакт\w*|ник|логин|аккаунт|переписываться|переписка|перейд\w*|перейти|давай|давайте|есть"
)
_B = rf"(?<![{_L}\d_])"
_E = rf"(?![{_L}\d_])"
_MENTION = re.compile(
    rf"{_B}(?:{_INTENT}){_E}(?:[^\w\n]+[\w-]+){{0,3}}?[^\w\n]+(?:{_MESSENGER}){_E}"
    rf"|{_B}(?:{_MESSENGER})\s*[:=]\s*[^\s]+"
    rf"|{_B}(?:{_MESSENGER})\s+[a-z][a-z0-9_.]{{3,}}",
    re.I,
)


@dataclass(frozen=True)
class Hit:
    kind: str
    start: int
    end: int

    def as_dict(self) -> dict:
        return {"kind": self.kind, "start": self.start, "end": self.end}


def _digits(token: str) -> str:
    t = token.lower()
    if t in _DIGIT_WORDS:
        return _DIGIT_WORDS[t]
    return t.replace("o", "0").replace("о", "0")


def _groups(text: str, start: int, end: int) -> list[tuple[str, int, int]]:
    """Группы цифр внутри найденного отрезка: подряд идущие символы-цифры склеиваются,
    цифра словом — всегда отдельная группа."""
    groups: list[tuple[str, int, int]] = []
    prev_end, prev_word = None, True
    for u in _UNIT_RE.finditer(text, start, end):
        tok = u.group(0)
        word = tok.lower() in _DIGIT_WORDS
        if groups and not word and not prev_word and prev_end == u.start():
            d, s0, _ = groups[-1]
            groups[-1] = (d + _digits(tok), s0, u.end())
        else:
            groups.append((_digits(tok), u.start(), u.end()))
        prev_end, prev_word = u.end(), word
    return groups


def _is_phone(digits: str, plus_start: bool, seps: str, single_group: bool) -> bool:
    n = len(digits)
    if plus_start:
        return 10 <= n <= 15
    if n == 11:
        return digits[0] in "78" and digits[1] in "3489"
    if n == 10:
        if digits[0] == "9":
            return True
        # городской номер без восьмёрки: 495 123-45-67, (812) 123 45 67
        return digits[0] in "348" and (single_group or bool(re.search(r"[-()]", seps)))
    return False


def _allowed_end(groups, i: int) -> int | None:
    digits = ""
    for j in range(i, len(groups)):
        digits += groups[j][0]
        if len(digits) > 11:
            return None
        if digits in ALLOWED_NUMBERS or (digits[:1] == "7" and "8" + digits[1:] in ALLOWED_NUMBERS):
            return j
    return None


def _phone_hits(text: str) -> list[Hit]:
    hits: list[Hit] = []
    for run in _RUN.finditer(text):
        raw = run.group(0).lower()
        plus = raw.startswith("+") or raw.startswith("плюс")
        groups = _groups(text, run.start(), run.end())
        i = 0
        while i < len(groups):
            allowed = _allowed_end(groups, i)
            if allowed is not None:
                i = allowed + 1
                continue
            digits, found = "", None
            for j in range(i, len(groups)):
                digits += groups[j][0]
                if len(digits) > 15:
                    break
                seps = text[groups[i][2]:groups[j][1]]
                if _is_phone(digits, i == 0 and plus, seps, i == j):
                    found = j
                    break
            if found is not None:
                start = run.start() if (i == 0 and plus) else groups[i][1]
                hits.append(Hit("phone", start, groups[found][2]))
                i = found + 1
            else:
                i += 1
    return hits


def find_contacts(text: str) -> list[Hit]:
    """Все найденные фрагменты, без пересечений, по порядку."""
    if not text:
        return []
    found: list[Hit] = []
    for kind, rx in (("email", _EMAIL), ("link", _LINK), ("handle", _HANDLE), ("link", _MENTION)):
        found += [Hit(kind, m.start(), m.end()) for m in rx.finditer(text)]
    found += _phone_hits(text)
    found.sort(key=lambda h: (h.start, -(h.end - h.start)))
    out: list[Hit] = []
    for h in found:
        if out and h.start < out[-1].end:
            continue
        out.append(h)
    return out


def has_contacts(text: str) -> bool:
    return bool(find_contacts(text))


def describe(hits: list[Hit]) -> str:
    kinds = []
    for h in hits:
        label = KIND_LABELS[h.kind]
        if label not in kinds:
            kinds.append(label)
    return ", ".join(kinds)
